Fix #5180: Fix race condition in class loading cycle detection

gzm0 opened the pull request:
The key insight here is that we need an intermediate loading state in which the infos are loaded but the class' parent chain is not linked yet.

We must wait for all the infos of an inheritance cycle to be loaded to detect the cycle reliably. However, we must not wait for the linking to be completed, otherwise we end up with a future-cycle.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala:
```diff
@@ -375,112 +375,152 @@ private class AnalyzerRun(config: CommonPhaseConfig, initial: Boolean,
+    def complete(result: LoadingResult): Unit =
+      promise.success(result)
+
+    def checkParentChain(): Future[Option[List[ClassName]]] =
+      checkParentChain(Set()).map(_.map(_.cycle))
```
This method deserves a comment about its result. What does the result mean?

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala:
```diff
@@ -495,7 +535,7 @@ private class AnalyzerRun(config: CommonPhaseConfig, initial: Boolean,
       val syntheticKind: Option[SyntheticClassKind],
       val nonExistent: Boolean)
-      extends Analysis.ClassInfo with ClassLoadingState with LoadingResult with ModuleUnit {
+      extends Analysis.ClassInfo with LoadingResult with InfoLoadedState with ModuleUnit {
```
This seems redundant. IIUC `LoadingResult extends InfoLoadedState`, so:
```suggestion
      extends Analysis.ClassInfo with LoadingResult with ModuleUnit {
```
?

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala:
```diff
@@ -375,112 +375,152 @@ private class AnalyzerRun(config: CommonPhaseConfig, initial: Boolean,
         syntheticKind: Option[SyntheticClassKind]): ClassLoadingState = {
-      var loading: LoadingClass = null
+      var loading: LoadingInfos = null
       val state = _classInfos.getOrElseUpdate(className, {
-        loading = new LoadingClass(className)
+        loading = new LoadingInfos
         loading
       })
```
Looking at this, I wonder if we need to mark `loading` as `@volatile`, to ensure safe publication of the assignment across threads? Or perhaps an `AtomicReference`?

sjrd replied:
> No, I don't think so. `loading` is only ever written to by the calling thread (at least for any reasonable implementation of `getOrElseUpdate`).

Right, indeed. Sounds good.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala:
```diff
@@ -375,112 +375,152 @@ private class AnalyzerRun(config: CommonPhaseConfig, initial: Boolean,
-            CycleInfo(className :: c, root)
+      // Start loading the parents.
+      val parentNames = data.superClass.toList ++ data.interfaces
+      val parentsFuture = Future.traverse(parentNames) {
+        ensureLoading(_, syntheticKind = None) match {
```
Consider naming the lambda argument here; it's a bit non-obvious:
```suggestion
      val parentsFuture = Future.traverse(parentNames) { parentName =>
        ensureLoading(parentName, syntheticKind = None) match {
```

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala:
```diff
@@ -375,112 +375,152 @@ private class AnalyzerRun(config: CommonPhaseConfig, initial: Boolean,
   }
 
-  private sealed trait LoadingResult
   private sealed trait ClassLoadingState
+  private sealed trait InfoLoadedState extends ClassLoadingState
+  private sealed trait LoadingResult extends InfoLoadedState
```
Perhaps consider adding methods on these types to resolve them to the possible tiers. IIUC the tiers we need are
```scala
  private sealed trait ClassLoadingState {
    def resolveToInfoLoadedState: Future[InfoLoadedState]
    def resolveToLoadingResult: Future[LoadingResult]
  }
```
This would lighten the notation inside the `traverse` calls. More importantly, IMO it would ease understanding that all cases are really covered.

(They could be implemented either in an OO fashion or with pattern matching.)

sjrd replied:
Thanks. I find it much better this way.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala:
```diff
@@ -375,112 +375,152 @@ private class AnalyzerRun(config: CommonPhaseConfig, initial: Boolean,
               Future.successful(createMissingClassInfo(className))
             }
-            info.flatMap { data =>
+            info.foreach { data =>
```
Do I interpret this comment right, that it is a regression you still intend to fix?

gzm0 replied:
Yes. Otherwise the tests time out. In fact, it points to a weakness in the implementation pattern overall: if `doLoad` throws, we have never completing futures. I think that needs fixing :-/

gzm0 commented:
I have updated this to not drop exceptions (or maybe rather `Failure`s) on the ground. It took quite some massaging to be able to do this safely. I have added an optional commit with more typing, but less boilerplate.

gzm0 commented:
Closing in favor of https://github.com/scala-js/scala-js/pull/5241
